### ISSUE TITLE
Read SID registers and have them on RAM

### ISSUE THREAD

**Issue by munshkr** [c64scene-ar/sid-player#1]
Looks like all SID registers (except for the 3rd voice) are write-only. Here is a solution I found on the [csdb forum](http://csdb.dk/forums/index.php?roomid=11&topicid=32097&firstpost=2):

> you have 2 options(others will bring up more:)
> 1. you turn on d000 ram, and call the player like that. then you can read the values from ram, and write them back to the sid regs afterwards. but this will distort the player's output. (mostly noticable to music people, or at specific players)
> 2. you recode all sta $d400 (generally any sid write) to jsr someroutine, where someroutine can store the value and then update the sid register. cubic player does this, it automatically recodes every player.

I think the second option is the safest. So we need a script that recodes every write to $d400-$d7ff on a .sid file into a _jsr_ that stores the value somewhere on RAM (to be defined) before.

**Comment by munshkr** [comment]
Or maybe we could use [SIDdump](http://cadaver.homeftp.net/tools/siddump.zip) by Cadaver :D 

```
Load address: $1000 Init address: $1000 Play address: $1003
Calling initroutine with subtune 0
Calling playroutine for 3000 frames, starting from frame 0
Middle C frequency is $1168

| Frame | Freq Note/Abs WF ADSR Pul | Freq Note/Abs WF ADSR Pul | Freq Note/Abs WF ADSR Pul | FCut RC Typ V |
+-------+---------------------------+---------------------------+---------------------------+---------------+
|     0 | 0000  ... ..  00 0000 000 | 0000  ... ..  00 0000 000 | 0000  ... ..  00 0000 000 | 0000 F0 Off 0 |
|     1 | 0116  ... ..  .. .... ... | 0116  ... ..  .. .... ... | 0116  ... ..  .. .... ... | .... .. ... F |
|     2 | ....  ... ..  40 0F80 ... | ....  ... ..  40 0F00 ... | ....  ... ..  .. .... ... | .... .. ... . |
|     3 | ....  ... ..  .. .... ... | ....  ... ..  .. .... ... | ....  ... ..  .. .... ... | .... .. ... . |
|     4 | ....  ... ..  09 0C04 800 | ....  ... ..  09 0C02 ... | ....  ... ..  .. .... ... | .... .. ... . |
|     5 | 45A0  C-6 C8  81 .... ... | 0343  G-1 93  41 .... 460 | ....  ... ..  .. .... ... | .... .. ... . |
|     6 | 07C1 (A#2 A2) 41 .... ... | ....  ... ..  .. .... 480 | ....  ... ..  .. .... ... | .... .. ... . |
|     7 | 06E8 (G#2 A0) .. .... ... | ....  ... ..  .. .... 4A0 | ....  ... ..  .. .... ... | .... .. ... . |
|     8 | 02BD (E-1 90) .. .... ... | ....  ... ..  .. .... 4C0 | ....  ... ..  .. .... ... | .... .. ... . |
|     9 | 0374 (G#1 94) .. .... ... | ....  ... ..  .. .... 4E0 | ....  ... ..  .. .... ... | .... .. ... . |
|    10 | 0116 (C-0 80) 10 .... ... | ....  ... ..  40 .... 500 | ....  ... ..  .. .... ... | .... .. ... . |
|    11 | ....  ... ..  .. .... ... | ....  ... ..  .. .... 520 | ....  ... ..  .. .... ... | .... .. ... . |
|    12 | ....  ... ..  .. .... ... | ....  ... ..  .. .... 540 | ....  ... ..  .. .... ... | .... .. ... . |
```

**Comment by munshkr** [comment]
Recoding every store with a JSR at first looks like the easiest way, but if we want to do it generally, we should do it for every possible store instruction (STA, STX, STY, etc) _and_ every SID register (there are like 32), a bit of a PITA.

Another way (which is what I'm actually trying now) is to insert a store to another place in RAM. For example, this:

``` asm
LDA $1577,X
STA $D401,Y
LDA $1598,X
STA $D406,Y
```

would turn into:

``` asm
LDA $1577,X
STA $C001,Y
STA $D401,Y
LDA $1598,X
STA $C006,Y
STA $D406,Y
```

where $C000 is a base address I'd specify to the recoder.

Now, because inserting another instruction changes the length of the containing subroutine (with a JSR this doesn't happen, because all these stores and a JSR both are 3 bytes long), we need to, for every store recoded, fix all subsequent label references (JMP with absolute addresses) by adding 3.

**Comment by munshkr** [comment]
OK, so I went back with the idea of using a JSR routine instead of inserting instructions, because it proved to be really hard, as I had to dissasembly the SID file and replace every label reference (absolute and relative), which is not trivial at all.

`recode_sid.py` searches for store instructions to SID registers and replaces them with a JSR to a specific routine that does the write to its SID register and somewhere else on RAM. Each subroutine is 7 bytes long (although I padded it to 8 bytes) and how many subroutines needs to be added depends on the .sid file.
